### PR TITLE
feat: Add support for HTTP load balancing between the proxy and upstream server groups

### PR DESCRIPTION
**Pull request by SchoNie**
Currently all connections are forwarded upstream with Round Robin algorithm. (default Nginx)
This pull request adds a container labels to specify the load balancing algorithm. Which can be used to have session persistence / affinity.
Some web applications which use the HTTP long-polling transport like websocket require sticky sessions.

Related #871, #299, #221, #886

Note if #1934 is merged: When using load balancing methods other than the default round-robin method, it is necessary to activate them before the keepalive directive.

If you think this PR is usefull I also plan to add the 'weight' directive.

I am trying to write a test. I have no pytest experience and having a little hard time to setup the test suite.
```
def test_web_loadbalance(docker_compose, nginxproxy):
    conf = nginxproxy.get_conf().decode('ASCII')
    r1 = nginxproxy.get("http://web.nginx-proxy.tld")
    r2 = nginxproxy.get("http://web.nginx-proxy.tld")
    assert r1.status_code == 200
    assert r2.text == r1.text
```
If I am correct the test webserver should output the container hostname.
But this always passes. In my first test when still using round robin I would expect this to fail.
Could someone point me in the right direction?

### CODE REVIEW

**Comment by rhansen** [comment]
> If you think this PR is useful

I do!  Thank you for the PR!

> If I am correct the test webserver should output the container hostname.

Yes.

> But this always passes. In my first test when still using round robin I would expect this to fail.

I too would expect that test to fail.  Some questions:

  * Could this be caused by `Connection: keep-alive` from the test client to nginx?
      * Does nginx use the same upstream server to handle all requests transmitted in the same persistent connection?
      * Does the HTTP client in the `nginxproxy` fixture use persistent connections?
  * What does the associated `.yml` file contain?
  * What do you get when you manually query? Like this:
    ```shell
    docker-compose -f test/test_load_balancing.yml up -d
    ip=$(docker container inspect -f '{{.NetworkSettings.Networks.bridge.IPAddress}}' test_sut_1)
    curl --connect-to ::${ip}: http://web.nginx-proxy.tld/
    ```
  * Does your nginx-proxy container have any extra config files that enable content caching?  (run `docker exec test_sut_1 nginx -T` to see nginx's full config)

**Comment by rhansen** [comment]
Given that the user provides the full line including the semicolon, maybe it would be better to just add an `include` directive if a certain file exists.  For example, look for `/etc/nginx/upstream.d/${upstream_name}.conf` first, and if that doesn't exist, look for `/etc/nginx/upstream.d/default.conf`.  If either exists, that file will be included.  Then users can put the load balancing directive in the include file.  What do you think?

**Comment by SchoNie** [comment]
Thank you Richard! Your hints pointed me in the right direction and I had a small error in my compose file.
Added some initial tests. All feedback is appreciated.

>  

> Given that the user provides the full line including the semicolon, maybe it would be better to just add an `include` directive if a certain file exists. For example, look for `/etc/nginx/upstream.d/${upstream_name}.conf` first, and if that doesn't exist, look for `/etc/nginx/upstream.d/default.conf`. If either exists, that file will be included. Then users can put the load balancing directive in the include file.
> 
> Two annoying things about my suggestion:
> 
> * The user has to be able to determine the upstream name that nginx-proxy generates, which I plan on changing ([feat: Use container names for upstream names to avoid repeats #1938](https://github.com/nginx-proxy/nginx-proxy/pull/1938)).
> * The user has to create a separate file and bind mount it into the container, rather than simply define a label.
> 
> Maybe a compromise would be to rename the label in this PR from `*.loadbalance` to something like `*.upstream_directives` and adjust the documentation to say that any directives suitable in an `upstream` block can be placed here, such as load balancing directives. One problem with this: what if multiple containers in the same upstream set this label? This PR picks an arbitrary winner; if the label is renamed, maybe it would be better to concatenate the values. There are usability downsides either way.

I was at first thinking to make the loadbalancing a true/false option. But then because of all flexible loadbalance options I decided to keep it clean and just insert the directive as-is from the label to prevent from a lot if/else stuff.
And focused more on writing the documentation with good examples.

I do like your suggestions. They gave me new ideas. My thoughts:
Including a upstream.d file is smart, but indeed your predicable name PR needs to be in there first. And maybe not so easy for the user to find out.
More and more apps use websockets and now we can provide a documented and unit-tested approach to handle this.

Renaming to `*.upstream_directives` is a possibility but for me it sounds a little to global. I don't see much more options in the [upstream module](http://nginx.org/en/docs/http/ngx_http_upstream_module.html) an average user would use. (The most interesting are related to the keepalive timeouts) and like you say can be dangerous if set different on multiple containers.

I rather would expand this PR to include the server weight and make a separate PR to include a upstream.d file. Which we can benefit from in more complex scenarios.

File: README.md [hunk 1]
```diff
@@ -373,6 +373,50 @@ docker run -d -p 80:80 -p 443:443 \
 
 You'll need apache2-utils on the machine where you plan to create the htpasswd file. Follow these [instructions](http://httpd.apache.org/docs/2.2/programs/htpasswd.html)
 
+### Upstream (Backend) Server HTTP Load Balancing Support
+
+> **Warning**
+> This feature is experimental.  The behavior may change (or the feature may be removed) without warning in a future release, even if the release is not a new major version.  If you use this feature, please provide feedback so we know whether it is working properly and can be promoted to officially supported.
+To enable HTTP load balancing between `nginx-proxy` and a group of backend servers, set the `com.github.nginx-proxy.nginx-proxy.loadbalance` label on the server's container to the load balancing algorithm which should be used. See the [nginx load balancing documentation](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-load-balancer/) and the [Docker label documentation](https://docs.docker.com/config/labels-custom-metadata/) for details.
+The load balancing algorithm is enabled for the containers sharing the same `VIRTUAL_HOST`
+#### Choosing a Load-Balancing Method
```
Comment (rhansen) [inline README.md:382]: I don't think it's necessary to go into detail about choosing a load balancing method.  I think it is sufficient to just link to the nginx documentation and explain how to set up load balancing in nginx-proxy.  Something like this:

> ### Upstream (Backend) Server HTTP Load Balancing Support
>
>> **Warning**
>> *(experimental warning goes here)*
>
> If you have multiple containers with the same `VIRTUAL_HOST` and `VIRTUAL_PATH` settings, nginx will spread the load across all of them.  To change the load balancing algorithm from nginx's default (round-robin), set the `com.github.nginx-proxy.nginx-proxy.loadbalance` label on one or more of your application containers to the desired load balancing directive.  See the [`ngx_http_upstream_module` documentation](https://nginx.org/en/docs/http/ngx_http_upstream_module.html) for available directives.
>
>> **Note**
>>   * Don't forget the terminating semicolon (`;`).
>>   * If you are using Docker Compose, remember to escape any dollar sign (`$`) characters (`$` becomes `$$`).
>
> Docker Compose example:
>
> ```yaml
> services:
>   nginx-proxy:
>     image: nginxproxy/nginx-proxy
>     ports:
>       - "80:80"
>     volumes:
>       - /var/run/docker.sock:/tmp/docker.sock:ro
>     environment:
>       HTTPS_METHOD: nohttps
>   myapp:
>     image: jwilder/whoami
>     expose:
>       - "8000"
>     environment:
>       VIRTUAL_HOST: myapp.example
>       VIRTUAL_PORT: "8000"
>     labels:
>       com.github.nginx-proxy.nginx-proxy.loadbalance: "hash $$remote_addr;"
>     deploy:
>       replicas: 4
> ```

File: nginx.tmpl [hunk 2]
```diff
@@ -198,6 +198,11 @@
 {{- define "upstream" }}
 upstream {{ .Upstream }} {
     {{- $server_found := false }}
+    {{- $loadbalance := first (keys (groupByLabel .Containers "com.github.nginx-proxy.nginx-proxy.loadbalance")) }}
+    {{- if $loadbalance }}
+    # Loadbalancing method: {{ $loadbalance }}
```
Comment (rhansen) [inline nginx.tmpl:203]: I don't think it's useful to repeat the value of the `$loadbalance` variable here.

```suggestion
    # From the container's loadbalance label:
```

File: test/test_loadbalancing.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
+  loadbalance-hash-1:
```
Comment (rhansen) [inline test/test_loadbalancing.yml:1]: Missing `services:` line?

```suggestion
services:
  loadbalance-hash-1:
```

File: test/test_loadbalancing.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    labels:
+      com.github.nginx-proxy.nginx-proxy.loadbalance: "hash $$remote_addr;"
+
+  loadbalance-hash-2:
+    image: web
+    expose:
+      - "82"
+    environment:
+      WEB_PORTS: 82
+      VIRTUAL_HOST: loadbalance-enabled.nginx-proxy.tld
```
Comment (rhansen) [inline test/test_loadbalancing.yml:17]: Rather than repeat the container, use `deploy.replicas`:

```suggestion
    deploy:
      replicas: 2
```

**Review by rhansen — CHANGES_REQUESTED** [review]

File: test/test_loadbalancing.yml (hunk 3 above)
Comment (SchoNie) [reply test/test_loadbalancing.yml:1]: It's missing in more tests. For example: [test_keepalive.yml](https://github.com/rhansen/nginx-proxy/blob/keepalive/test/test_keepalive.yml) and [test_log_format.yml](https://github.com/nginx-proxy/nginx-proxy/blob/main/test/test_log_format.yml) and more.
So I was not sure what is prefered.

Comment (rhansen) [reply test/test_loadbalancing.yml]: I should fix those.  Omitting it means use the old v1 semantics, which doesn't matter for those tests, but they should be updated anyway for consistency.  It is needed here in order to use `deploy.replicas`.

**Review by rhansen — APPROVED** [review]
LGTM.  What do you think @buchdag?

We should remember to squash the commits when merging (or you can squash them yourself @SchoNie and force push before we merge).

File: README.md [hunk 5]
```diff
@@ -373,6 +373,42 @@ docker run -d -p 80:80 -p 443:443 \
 
 You'll need apache2-utils on the machine where you plan to create the htpasswd file. Follow these [instructions](http://httpd.apache.org/docs/2.2/programs/htpasswd.html)
 
+### Upstream (Backend) Server HTTP Load Balancing Support
+
+> **Warning**
+> This feature is experimental.  The behavior may change (or the feature may be removed) without warning in a future release, even if the release is not a new major version.  If you use this feature, please provide feedback so we know whether it is working properly and can be promoted to officially supported.
```
Comment (buchdag) [inline README.md:379]: I'll have the same request as on https://github.com/nginx-proxy/nginx-proxy/pull/1934 : please open a discussion for feedbacks on this feature first, then include a link to it in the documentation.

**Review by buchdag — CHANGES_REQUESTED** [review]

**Comment by buchdag** [comment]
@rhansen apart from the request change related to the feedback discussion and it inclusion in the docs, this LGTM (and yes we should squash the commits).

File: README.md (hunk 5 above)
Comment (SchoNie) [reply README.md]: Done. #2195 
And also squashed.

**Review by rhansen — DISMISSED** [review]

**Review by buchdag — APPROVED** [review]

**Pull request merged**